# Day 12 Assignment : 
Apply the Parametric and Non-Parametric tests on the Attrition Analysis Dataset and provide the inference.

### Step 1 : 
Importing the libraries and dataset 

In [6]:
#importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [7]:
dt = pd.read_csv("general_data.csv")
dt.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

## Step 2 : 
Data Preprocessing Stage to remove the null and the duplicate values

In [8]:
dt.isnull()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  \
0     False      False           False       False             False   
1     False      False           False       False             False   
2     False      False           False       False             False   
3     False      False           False       False             False   
4     False      False           False       False             False   
...     ...        ...             ...         ...               ...   
4405  False      False           False       False             False   
4406  False      False           False       False             False   
4407  False      False           False       False             False   
4408  False      False           False       False             False   
4409  False      False           False       False             False   

      Education  EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0         False           False          False       False   False  ...   
1         False           False          False       False   False  ...   
2         False           False          False       False   False  ...   
3         False           False          False       False   False  ...   
4         False           False          False       False   False  ...   
...         ...             ...            ...         ...     ...  ...   
4405      False           False          False       False   False  ...   
4406      False           False          False       False   False  ...   
4407      False           False          False       False   False  ...   
4408      False           False          False       False   False  ...   
4409      False           False          False       False   False  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
0                  False   False              False          False   
1                  False   False              False          False   
2                  False   False              False          False   
3                  False   False              False          False   
4                  False   False              False          False   
...                  ...     ...                ...            ...   
4405               False   False              False          False   
4406               False   False              False          False   
4407               False   False              False          False   
4408               False   False              False          False   
4409               False   False              False          False   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                False              False                  False   
1                False              False                  False   
2                False              False                  False   
3                False              False                  False   
4                False              False                  False   
...                ...                ...                    ...   
4405             False              False                  False   
4406             False              False                  False   
4407             False              False                  False   
4408             False              False                  False   
4409             False               True                  False   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0              False                    False                 False  
1              False                    False                 False  
2              False                    False                 False  
3              False                    False                 False  
4              False                    False                 False  
...              ...                      ...                   ...  
4405           False                    False                 False  
4406           False    

In [9]:
# We check for duplicate values
dt.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 4410, dtype: bool

In [13]:
# We remove the null and the duplicate values
dt1 = dt.dropna()
dt2 = dt1.drop_duplicates()
dt2

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
1      31       Yes  Travel_Frequently  Research & Development   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4404   29        No      Travel_Rarely                   Sales   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
1                   10          1  Life Sciences              1           2   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
...                ...        ...            ...            ...         ...   
4404                 4          3          Other              1        4405   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
1     Female  ...                 0.0      Y                23              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
...      ...  ...                 ...    ...               ...            ...   
4404  Female  ...                 1.0      Y                18              8   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
1                    1               6.0                      3   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
...                ...               ...                    ...   
4404                 0               6.0                      2   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
1                  5                        1                     4  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
...              ...                      ...                   ..

## Step 3 : 
Now we create two separate datasets for Attrition YES and NO to be able to proceed with the tests

In [28]:
dt_Ayes = dt2[dt2["Attrition"]=="Yes"]
dt_Ayes

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                   10          1  Life Sciences              1           2   
6                   11          2        Medical              1           7   
13                   1          1        Medical              1          14   
28                   1          2        Medical              1          29   
30                   4          3        Medical              1          31   
...                ...        ...            ...            ...         ...   
4381                 7          1  Life Sciences              1        4382   
4386                11          4      Marketing              1        4387   
4388                 1          3  Life Sciences              1        4389   
4391                23          1  Life Sciences              1        4392   
4402                 2          3      Marketing              1        4403   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1     Female  ...                 0.0      Y                23              8   
6       Male  ...                 2.0      Y                20              8   
13      Male  ...                 1.0      Y                11              8   
28      Male  ...                 3.0      Y                14              8   
30      Male  ...                 2.0      Y                11              8   
...      ...  ...                 ...    ...               ...            ...   
4381  Female  ...                 1.0      Y                21              8   
4386    Male  ...                 5.0      Y                21              8   
4388    Male  ...                 7.0      Y                11              8   
4391    Male  ...                 0.0      Y                11              8   
4402    Male  ...                 6.0      Y                11              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
1                    1               6.0                      3   
6                    1               5.0                      2   
13                   2              10.0                      4   
28                   1              19.0                      2   
30                   0               5.0                      5   
...                ...               ...                    ...   
4381                 0               4.0                      2   
4386                 0               8.0                      2   
4388                 0              13.0                      2   
4391                 0               4.0                      2   
4402                 1              17.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1                  5                        1                     4  
6                  0                        0                     0  
13                10                        9                     9  
28                 1                        0                     0  
30                 3                        0                     2  
...              ...                      ...                   ..

In [29]:
dt_ANo = dt2[dt2["Attrition"]=="No"]
dt_ANo

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
5      46        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4404   29        No      Travel_Rarely                   Sales   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
5                    8          3  Life Sciences              1           6   
...                ...        ...            ...            ...         ...   
4404                 4          3          Other              1        4405   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
5     Female  ...                 3.0      Y                13              8   
...      ...  ...                 ...    ...               ...            ...   
4404  Female  ...                 1.0      Y                18              8   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
5                    0              28.0                      5   
...                ...               ...                    ...   
4404                 0               6.0                      2   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
5                  7                        7                     7  
...              ...                      ...                   ..

## Step 4 : Parametric Tests 

##### 1 . ***Mann Whiteny Test*** of Attrition with the different Variables :
Format for all Hypothesis Test : 

**H0** - There is no significant difference between Attrition_Yes and Attrition_No datasets for {VARIABLE}

**H1** - There is a significant difference between Attrition_Yes and Attrition_No datasets for {VARIABLE}

In [31]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.DistanceFromHome, dt_ANo.DistanceFromHome) 
stats, p

(1295261.0, 0.488538986087403)

As we can see the p-value `0.5317715668019634>0.05` so we accept the null hypothesis.
So H0 is accepted and H1 is rejected

In [32]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.Age, dt_ANo.Age) 
stats, p

(949178.0, 7.98668614365882e-30)

As we can see the p-value ` 7.98668614365882e-30<0.05` so we reject the null hypothesis.
Therefore there is a significant difference between ***Attrition_yes*** and ***Attrition_No*** for ***Age*** 

In [33]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.Education, dt_ANo.Education) 
stats, p

(1258277.0, 0.09888570408770686)

As we can see the p-value `0.09888570408770686>0.05` so we accept the null hypothesis.
Therefore there is no significant difference between ***Attrition_yes*** and ***Attrition_No*** for ***Education*** 

In [34]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.NumCompaniesWorked, dt_ANo.NumCompaniesWorked) 
stats, p

(1238814.5, 0.02793197853866981)

As we can see the p-value `<0.05` so we reject the null hypothesis.
Therefore there is no significant difference between ***Attrition_yes*** and ***Attrition_No*** for ***NumCompaniesWorked***  

In [35]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.PercentSalaryHike, dt_ANo.PercentSalaryHike) 
stats, p

(1231873.5, 0.017810794960084964)

As we can see the p-value `<0.05` so we reject the null hypothesis.
Therefore there is no significant difference between ***Attrition_yes*** and ***Attrition_No*** for ***PercentSalaryHike***  

In [36]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.TotalWorkingYears, dt_ANo.TotalWorkingYears) 
stats, p

(895173.5, 2.741211827689903e-39)

As we can see the p-value `2.741211827689903e-39<<<0.05` so we reject the null hypothesis.
Therefore there is a significantdifference between ***Attrition*** and ***TotalWorkingYears*** 

In [37]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.YearsAtCompany, dt_ANo.YearsAtCompany) 
stats, p

(912579.0, 3.3433144809752036e-36)

As we can see the p-value `3.3433144809752036e-36<0.05` so we reject the null hypothesis.
Therefore there is a significant difference between ***Attrition_yes and Attrition_No*** for ***YearsAtCompany*** 

In [38]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.YearsSinceLastPromotion, dt_ANo.YearsSinceLastPromotion) 
stats, p

(1196606.0, 0.00037904698157957496)

As we can see the p-value `0.00037904698157957496<0.05` so we reject the null hypothesis.
Therefore there is a significant correaltion between ***Attrition_Yes*** and ***Attrition_No** for ***YearsSinceLastPromotion*** 

In [40]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.YearsWithCurrManager, dt_ANo.YearsWithCurrManager) 
stats, p

(945958.5, 5.420302388722274e-31)

As we can see the p-value `5.420302388722274e-31<0.05` so we reject the null hypothesis.
Therefore there is a significant difference between ***Attrition_Yes*** and ***Attrition_No** for ***YearsWithCurrentManager*** 

In [41]:
from scipy.stats import mannwhitneyu
stats,p = mannwhitneyu(dt_Ayes.TrainingTimesLastYear, dt_ANo.TrainingTimesLastYear) 
stats, p

(1225582.0, 0.008107344081224082)

 we can see the p-value `0.00037904698157957496<0.05` so we reject the null hypothesis.
Therefore there is a significant difference between ***Attrition_Yes*** and ***Attrition_No** for ***TrainingTimeLastYear*** 